### DISCUSSION TITLE
How do people feel about the use of unicode in Oceananigans' API?

### DISCUSSION THREAD

**Discussion by glwagner** [CliMA/Oceananigans.jl#2303] [Ideas]
On slack we recently had a conversation with @robertwhelber and @aramirezreyes about the downsides of unicode in our API. There are really two issues:

1. Some unicode symbols are not supported by some fonts. When a font does not support a certain symbol, readers see `⍰` instead. There's even a [julia-specific font called JuliaMono](https://juliamono.netlify.app/) that has "reasonable Unicode support" to at least partly solve this problem.

2. Not all text editors support Unicode entry. For example, vim requires a plugin (I use [Vundle](https://github.com/VundleVim/Vundle.vim) with `Plugin 'JuliaEditorSupport/julia-vim'`). @robertwhelber also reported a problem with VSCode on his machine.

This two issues mean that for some users, getting started with Oceananigans requires more than just installing Julia and typing `pkg> add Oceananigans`.

I think replacing Unicode in the source code is not only a gargantuan task but would also severely impact readability of the source code. We make widespread use of Unicode to produce easily readable, distinguishable operators like `δxᶜᵃᵃ, ℑxᶠᵃᵃ, ∂xᶜᶜᶠ` and we feel that unicode has solved some tough issues (eg making sense of the combinatorial explosion of operators required for staggered grids).

However, it may be possible to

1. Reduce unicode in our API somewhat, for example `Simulation.dt` instead of `Simulation.Δt`
2. Provide a "`Unicodeless`" API somehow... `nu` rather than `ν`, `dx` instead of `∂x`, `phi, lambda` instead of `φ λ`?

The problem with 1 is that it's just a half-measure that might reduce the quality of our API without really solving anyone's problem.
As @simone-silvestri said

> You're going to encounter Unicode eventually if you use Oceananigans

As for 2, it would require work and maintainence. However it's possible that we (or even someone else) could support option 2 as an external package, by redefining some key constructors and functions and re-exporting the already-acceptable versions.

My personal opinion that that using Unicode is important in making "code look like math" and writing "scripts that look like papers", and therefore a cornerstone in achieving one of our missions. However, I think we're also going to fail at achieving anything if lots of people are turned away at the door by Unicode.

**Comment by iuryt** [comment]
From my perspective, using Unicode is important. The code will look more like math and physics expressions, which makes easier to understand and convince people to use and adapt it, even if they are not very used to program in Julia.

**Comment by francispoulin** [comment]
I agree that using Unicode is very nice and hope we keep on doing it.  This unfortuantely means that some problems will arise. Maybe we can suggest a default editor in case people have problems that is more fool proof than others?

**Reply by glwagner** [reply]
Probably VSCode is the recommended editor. Docs look decent too https://www.julia-vscode.org/docs/dev/gettingstarted/

**Reply by francispoulin** [reply]
Sadly, I could not get VSCode to do Unicode and have since gone the way of atom.  I did post an issue but did not get a response.  

But I hope that I am the outlier and that VSCode will work for others.

**Reply by navidcy** [reply]
> But I hope that I am the outlier and that VSCode will work for others.

Works for me

**Reply by robertwhelber** [reply]
If I follow just the instructions here: https://www.julia-vscode.org/docs/dev/gettingstarted/
will Unicode characters also work?  Or, will I need to install other VSCode packages to get that working?
I appreciate tutorials like this, thanks.

**Reply by glwagner** [reply]
~~I'd think so but I don't know for sure. @navidcy can you confirm?~~ It's mentioned in the changelog: https://www.julia-vscode.org/docs/dev/changelog/#[0.14] a while ago now. Unicode / tab completion does work for me in VSCode (though it's a little slower than vim) so that's consistent with my experience.

Nothing is mentioned here: https://www.julia-vscode.org/docs/dev/userguide/editingcode/

I think a request for some information about Unicode and backslash completion in those docs might be appropriate for an issue here: https://github.com/julia-vscode/docs, though the top of the docs does mention they are currently being revamped so that might be in the works somehow.

**Comment by MasonProtter** [comment]
There's nothing wrong with providing a Unicode API. Unicode is great. It is quite problematic though to provide APIs that **require** unicode. 

There should always be ugly ASCII fallbacks.

**Reply by glwagner** [reply]
I wrote below:

I'm worried that it's not feasible / too much effort to maintain "two APIs". I second grave issue is that this produces "two languages"; ie users may have trouble reading scripts written using the alternate API that they don't use. I think we'll have to choose one.

**Comment by MilesCranmer** [comment]
Hey all,

From an outsider's perspective who just recently started trying out Oceananigans.jl and other related packages (really enjoying them overall!), the use of unicode was the biggest annoyance for me. Thus I would highly recommend option (2) above. For the record, this is the only Julia package I've encountered which uses a unicode API. I think unicode in scripts is fine, but for APIs it is too subjective to enforce.

While there are plugins I can install for IDEs to support it, there are so many things which break:

1. GitHub search/VSCode search/vim search. Most plugins don't help with this: try to search for a unicode symbol with the GitHub search console - tab completion does not work. If I want to search for occurrences of a particular function/symbol, I need to copy-paste the symbol which is (when compounded) a lot of unneeded effort.
2. github editor/github.dev/nano/other minimal editors which do not have any Julia plugins do not support unicode. (Sometimes I just want to make an edit on my phone...)
3. Bash. If I want to use `grep`, I need to know how to write the symbol in bash, or, again, copy-paste it to my terminal (assuming my terminal supports unicode).
4. IDE movement commands. vim's `f` command is very useful for navigation (see [this](https://stackoverflow.com/questions/12495442/what-do-the-f-and-t-commands-do-in-vim)), but I can't use it to jump to a unicode key. I also use [vim-sneak](https://github.com/justinmk/vim-sneak) for quickly moving around code, but it is also incompatible with unicode.
6. More keystrokes, especially for super/subscripts.

The argument I've seen in favor of unicode is that it looks better, but I can't see this either. _LaTeX_ math looks good, but unicode math looks terrible (subjective opinion!). It would be great if `_psi` was rendered as full-blown LaTeX $\psi$ - then I could use ascii to navigate, and see the LaTeX rendering when reading. But unicode math seems like the worst of both worlds... It also makes it so much harder to read. e.g., `δxᶜᵃᵃ` - I have to increase the zoom setting on my window (which is calibrated for normal fontsize) just to read the full variable name.

Anyways, this is all subjective. I think the point is that some users (like myself 🙂) are very anti-unicode - and thus an API should not enforce its use.

Again, awesome package! Happy either way you choose to go.
Best,
Miles

**Reply by glwagner** [reply]
This is good feedback.

As for option 2, I'm worried that it's not feasible / too much effort to maintain "two APIs". I second grave issue is that this produces "two languages"; ie users may have trouble reading scripts written using the alternate API that they don't use. I think we'll have to choose one.

`δxᶜᵃᵃ` is supposed to be internal, not user facing. Why do you find yourself having to read this function? (I'm curious because usually things like this means we are missing something in our public API.) I think `δxᶜᵃᵃ` has been very successful as internal developer notation, so I don't think we want to change that.

**Reply by MilesCranmer** [reply]
> As for option 2, I'm worried that it's not feasible / too much effort to maintain "two APIs". I second grave issue is that this produces "two languages"; ie users may have trouble reading scripts written using the alternate API that they don't use. I think we'll have to choose one.

I wonder if it is possible to build a macro for this? e.g.,

```julia
@deunicode function f(; ν)
    ν ^ 2
end
```

would `@macroexpand`, using a dictionary of unicode->ascii, to:

```julia
function f(; ν=nothing, nu=nothing)
    ν === nothing && nu === nothing && error("Provide one of either ν or nu.")
    ν !== nothing && nu !== nothing && error("Only provide one of ν or nu.")
    ν = ν === nothing ? nu : ν

    ν ^ 2
end
```

But if that is too tricky I guess (1) is a good option!

> `δxᶜᵃᵃ` is supposed to be internal, not user facing. Why do you find yourself having to read this function? (I'm curious because usually things like this means we are missing something in our public API.)

Sorry, I was just referencing the example from your initial post. The only public API with unicode I've encountered so far is `Δt` as a keyword argument, and and `TwoDNavierStokes.set_ζ!` (and `vars.ζ`) from GeophysicalFlows.jl (I can't really tell what symbol that is from my fonts - maybe zeta?). But I don't mind what others use for their internal API.

Aside: why use that, instead of something like `dx_caa`? The unicode version requires 21 keystrokes and the ascii version is just 6. I would argue the ascii version is even more readable since it doesn't require zooming in. Or `delta_`, `partial_`, etc., for the different symbols? (I can barely see the difference between `δ` and `∂` on my monitor... though I guess my browser fonts aren't good...)

**Reply by navidcy** [reply]
Regarding fonts, this is a good one: https://juliamono.netlify.app

**Reply by navidcy** [reply]
`TwoDNavierStokes.set_ζ!` yeah, that's a zeta... :( sorry it gives you trouble...

growing up with the greek alphabet perhaps this feels more natural to me than others

**Reply by MilesCranmer** [reply]
Maybe my eyes aren't good... I still find it so hard to read the examples on that page. e.g.,
<img width="151" alt="Screen Shot 2022-11-22 at 2 50 59 PM" src="https://user-images.githubusercontent.com/7593028/203408260-b618207c-3140-4626-ad46-ec3d751f992f.png">
1. Not sure if this is `e`, or a special type of unicode `e`.
2. Not sure if this is a special version of `I` or not.
3. The `l` is so small I can't make out if it's a `1`.

<img width="289" alt="Screen Shot 2022-11-22 at 2 51 08 PM" src="https://user-images.githubusercontent.com/7593028/203408578-babc4be3-7f14-48c7-8bb1-140e241f744f.png">

This one is really tricky. Skimming the page I thought there might be a capital `A`, but it's a `\Lambda`. I also still can't figure out if the subscript to the `u` is an `x` or a `\cross`.

The symbol `ʹ` is difficult. I can't tell if it's a transpose, or a variable name.

I guess once there is any unicode in the name space, then it makes hard to read everything else, since I'm never sure if it's the same symbol as is on my keyboard, or some complex unicode version...

I feel like the overall best thing to do would be to simply have rendering and code be two separate displays, rather than intermingled with unicode versions. But I guess this is quite subjective on my part.

**Reply by MilesCranmer** [reply]
> growing up with the greek alphabet perhaps this feels more natural to me than others

No worries! Understandable :)

**Reply by MilesCranmer** [reply]
Sometimes even ascii feels too flexible since the symbols I and l look so similar... Unicode just makes the problem 10x larger.

**Reply by MasonProtter** [reply]
For what it's worth, I think at least for user facing APIs, it shouldn't be very burdensome to maintain ASCII variants, certainly less work than "two APIs". All you need to do is write things like
```julia
export set_zeta!, set_ζ!

function set_zeta!(...)
    ...
end
const set_zeta! = set_ζ!
```
For internals yeah, you shouldn't worry about needing to provide ASCII stuff, but I think at least for names which appear in the user facing API, setting some `const` aliases for the various unicode things shouldn't be too hard. 

Maybe the right way to automate this is by making an `@export_with_ascii` macro such that if someone wrote
```julia
@export_with_ascii foo, set_ζ!, bar,  Δx₁
```
this gets transformed to
```julia
const set_zeta! =  set_ζ!
@doc "set_zeta! is an ascii alias for  set_ζ!" set_zeta! 

const Deltax_1 = Δx₁
@doc "Deltax_1 is an ascii alias for  Δx₁" Δx₁

export foo, set_ζ!, set_zeta!, bar, Δx₁, Deltax_1
```
or something.

**Reply by MasonProtter** [reply]
Personally, I'm a big unicode fan, and I really do think that unicode should be used and most complains about unicode in julia are actually complaints about the insufficient tools the person is using, not actually about unicode text itself. 

But even so I think at least for packages with many users, having an ASCII fallback is usually a good idea.

**Reply by glwagner** [reply]
Are there any other packages that provide "ASCII fallbacks"?

**Reply by MasonProtter** [reply]
@glwagner  To be honest, I had thought this would be much more common, but it turns out that a lot of packages just don't have Unicode APIs which is kidna too bad. Some examples I could find were https://github.com/JuliaMath/TensorCore.jl,  https://juliadiff.org/TaylorSeries.jl/latest/api/#TaylorSeries.gradient, and https://github.com/JuliaPhysics/Measurements.jl which has `a ± b` meaning `measurement(a, b)`.

**Reply by MilesCranmer** [reply]
I also haven't seen any other unicode-only APIs; this package was a first. Of the ones @MasonProtter shared it looks like they all manually define their ascii fallbacks; e.g., `measurement` (as Mason mentioned) for Measurements.jl; `hadamard(A, B) == A ⊙ B` for TensorCore.jl; and `gradient(f) == ∇(f)` for TaylorSeries.jl.

I do agree with Mason's comment about tools being insufficient is the main driver for slow adoption. I think once it's more universal (unfortunately this seems many years away...), then I agree it will be fine to have unicode-only APIs.

It would be great if all tools had unicode equivalents or if there was an easy workaround, but often one can't even make such a choice. Even the platform we are talking on right now does not have built-in unicode support in the search bar or comments.

In academia this is not as much of an issue since you can install whatever tool you like, but in (some) corporate software environments, they literally will not let you install custom plugins without it passing a long legal/security review - and even after, the plugin might require some parts of it be disabled. So requiring a custom plugin just to write an API call is a huge barrier to some users.

**Reply by glwagner** [reply]
> 4\. vim's `f` command is very useful for navigation (see [this](https://stackoverflow.com/questions/12495442/what-do-the-f-and-t-commands-do-in-vim)), but I can't use it to jump to a unicode key. I also use [vim-sneak](https://github.com/justinmk/vim-sneak) for quickly moving around code, but it is also incompatible with unicode.

Is this true even when using https://github.com/JuliaEditorSupport/julia-vim ?

**Reply by glwagner** [reply]
Thanks for doing that research @MasonProtter.

Up to this point we have prioritized the "readability" of user scripts. Our philosophy is that a user script should be as readable as a paper; therefore we would users to learn unicode so that they write readable scripts meant to be openly shared with an academic community. Similarly, to write an academic paper we're obligated to learn to type math into a written document.

I think this thread makes a convincing case that, if our priorities were different --- ie (i) attracting as wide a user base as possible and (ii) minimizing start up costs against other considerations --- we should change a substantial swath of user-facing code to eliminate unicode. So, we have to decide whether such a shift in priorities makes sense.

I'd like to point out that there is a separate concern: whether or not we should verbose notation or concise mathematical notation in the API. This is currently mixed; for example we can write

```julia
equation_of_state = LinearEquationOfState(thermal_expansion=2e-4)
```

reflecting a verbose API, or

```julia
closure = ScalarDiffusivity(ν=1.05e-6)
coriolis = FPlane(f=1e-4)
```

which uses mathematical notation. If we prioritize a verbose scheme then we might change the names of properties in many structs to things like

```julia
ScalarDiffusivity.viscosity
Simulation.time_step
FPlane.planetary_vorticity
```

etc. Switching to verbose names would eliminate a lot of unicode, though perhaps not all...

**Reply by MilesCranmer** [reply]
> Is this true even when using https://github.com/JuliaEditorSupport/julia-vim ?

Cool, thanks! Indeed that makes `f` work correctly in pure vim. (Although it seems like vim-sneak still doesn't work).

Unfortunately I normally use vim emulation within vscode which is incompatible with julia-vim, so `f` still won't work... https://marketplace.visualstudio.com/items?itemName=vscodevim.vim#emulated-plugins

Also, it seems like the Julia vscode doesn't patch the search bar - e.g.,

![Screen Shot 2022-11-24 at 6 57 08 PM](https://user-images.githubusercontent.com/7593028/203875998-ade4f4a1-6335-46aa-89f3-ca4a46df1216.png)

Then, hitting `<tab>` moves to the next option in the menu.

![Screen Shot 2022-11-24 at 6 57 12 PM](https://user-images.githubusercontent.com/7593028/203876004-c074cba6-bf53-4286-986e-d8446e019985.png)

It would be nice when unicode support is more common, but unfortunately the average user's tools just aren't ready yet.

**Reply by MilesCranmer** [reply]
> Up to this point we have prioritized the "readability" of user scripts. Our philosophy is that a user script should be as readable as a paper; therefore we would users to learn unicode so that they write readable scripts meant to be openly shared with an academic community.

I think unicode just isn't expressive enough for this, so it ends up being an awkward middle ground which is more confusing than if code was written normally, and then annotated with fully rendered LaTeX in a Jupyter notebook.

Take for example this snippet:

```julia
u, v, w = model.velocities
ζ = Field(∂x(v) - ∂y(u))
compute!(ζ)
```

If I pretend I am a naive user, I might look at this and be unsure if `∂x(v)` here means $\frac{\partial x}{\partial v}$ or $\frac{\partial v}{\partial x}$ or $d(x v)$ (i.e., the full derivative). It would be great if unicode had a character for $\frac{\partial}{\partial x}$, but it unfortunately does not, so I think the only way to be clear is to use something like

```julia
u, v, w = model.velocities
# $$\zeta = \frac{\partial v}{\partial x} - \frac{\partial u}{\partial y}$$
ζ = Field(∂(v, direction=:x) - ∂(u, direction=:y))
compute!(ζ)
```

and either write that comment in a jupyter notebook markdown cell, or use a TeX renderer like https://marketplace.visualstudio.com/items?itemName=vs-publisher-1305558.VsTeXCommentsExtension or https://orgmode.org/manual/LaTeX-fragments.html.

Being new to Julia myself, sometimes I get confused whether unicode is a function or part of a variable name. For example, if someone writes `∇f` or `f'`, those characters are actually functions – gradient and transpose, respectively – and not variable names. But with the current API, the code `grid.Δzᵃᵃᶜ` does not mean `grid.Δ(z)[:a, :a, :c]`, i.e., the laplacian of a field `z` indexed at `(:a, :a, :c)`, which it could very well be, but instead an actual variable named called `Δzᵃᵃᶜ`. So there's this ambiguity here which is confusing unless you are already comfortable with the code.

**Reply by glwagner** [reply]
It'd be interesting to hear if other users think that our API is awkward and confusing. If so, then we certainly have work to do.

Our intention is for users to read `∂x` as the operator

$$ \partial_x $$

so that `∂x(v) - ∂y(u)` is

$$ \partial_x v - \partial_y u $$

which is the notation for partial derivatives that most of the developers prefer (I think). We opted against sub/super scripts because they can be difficult to read, as noted, and don't seem necessary here.

`grid.Δzᵃᵃᶜ` isn't part of the user API. We are currently missing an API for grid spacings, areas, and volumes (though we have one for `nodes`). We should use the syntax `Center()` and `Face()` to refer to locations when we get around to designing this interface, similar to how we have

```julia
z = znodes(Center(), Center(), Center(), grid)
```

**Reply by glwagner** [reply]
> Also, it seems like the Julia vscode doesn't patch the search bar - e.g.,

This could be something to raise over at https://github.com/julia-vscode/julia-vscode

**Reply by tomchor** [reply]
> > Up to this point we have prioritized the "readability" of user scripts. Our philosophy is that a user script should be as readable as a paper; therefore we would users to learn unicode so that they write readable scripts meant to be openly shared with an academic community.
> 
> I think unicode just isn't expressive enough for this, so it ends up being an awkward middle ground which is more confusing than if code was written normally, and then annotated with fully rendered LaTeX in a Jupyter notebook.

@MilesCranmer you're raising some very interesting points, so thanks for the discussion! That said, I definitely have the opposite opinion here.

I remember when I first started with Oceananigans, part of what I liked was that the code (and the math in particular) was really easy and intuitive to read. So while I agree that unicode isn't perfect, it's way better than pure ascii code. LaTeX code rendered in Jupyter is definitely more legible, but that also requires extra work (and extra line for each expression). I never use Jupyter notebooks, for example, and work purely out of Vim. I don't know if there's even a way that I could render LaTeX comments there.


> Being new to Julia myself, sometimes I get confused whether unicode is a function or part of a variable name. For example, if someone writes `∇f` or `f'`, those characters are actually functions – gradient and transpose, respectively – and not variable names. But with the current API, the code `grid.Δzᵃᵃᶜ` does not mean `grid.Δ(z)[:a, :a, :c]`, i.e., the laplacian of a field `z` indexed at `(:a, :a, :c)`, which it could very well be, but instead an actual variable named called `Δzᵃᵃᶜ`. So there's this ambiguity here which is confusing unless you are already comfortable with the code.

It's my impression that this isn't a characteristic of Oceananigans per so, but of Julia as a whole. It seems to me that Julia gives users and developers a lot of freedom with characters and names, and other packages take a advantage of that freedom too, just like Oceananigans does. I think in the end lots of it is subjective, but for me personally it just took a bit of getting used to and now it feels as natural as python syntax (but more readable :) )

**Reply by MilesCranmer** [reply]
> It'd be interesting to hear if other users think that our API is awkward and confusing.

Sorry, that's not what I meant at all! I actually find the Oceananigans.jl API to be by far the most intuitive one for setting up hydro simulations I have seen out there. (I've even [tweeted](https://twitter.com/MilesCranmer/status/1584349390939967488) about how much I love the API!). But if there is any weakness to improve on, I think unicode is it.

I am trying to say that I am doubtful about the statement "a user script should be as readable as a paper" - I don't think unicode is expressive enough to do this, and trying to write out full equations using unicode seems like it will lead to confusion versus using full function names with annotations. I do find some of the examples in the docs fairly difficult to parse because of this. e.g., in this [example](https://clima.github.io/OceananigansDocumentation/stable/generated/shallow_water_Bickley_jet/), I can't tell what superscript or bar is just part of the variable name, or an operator: 

![Screen Shot 2022-11-26 at 5 27 25 PM](https://user-images.githubusercontent.com/7593028/204111169-03bb3d7f-0312-4d04-b5cd-466f638fe5c3.png)

I can still figure it out, but it's just a higher mental load to do so. I suppose eventually this would become intuitive, but it certainly seems like a steeper learning curve. (I guess this note is about the examples; not the API itself)

Luckily the current Oceananigans.jl API doesn't rely heavily on unicode, which is good! Although I do think having non-unicode versions of existing unicode (in API + examples) would be even better. (So 👍 for *We should use the syntax Center() and Face() to refer to locations when we get around to designing this interface*). Btw, `grid.Δzᵃᵃᶜ` appears in the Ocean mixing [example](https://clima.github.io/OceananigansDocumentation/stable/generated/ocean_wind_mixing_and_convection/#The-grid) - not sure if there's a user API equivalent or if it's just demo'ing some internals.

Regarding `∂x(v) - ∂y(u)` in particular, I can see why others like it, but I must confess I am really not a fan of this style because to my eyes it is ambiguous - both for the reasons noted earlier and because `∇x(v)` in Julia would typically mean `gradient(x)(v)` in autodiff packages, rather than a directional `gradient_x`. Maybe I will get used to it eventually but with fresh new-user eyes, I just can't say I found it easy to understand at first pass, though. Whereas the rest of the API, with full ASCII names, is super easy to understand and non-ambiguous.

> @tomchor: It's my impression that this isn't a characteristic of Oceananigans per so, but of Julia as a whole. It seems to me that Julia gives users and developers a lot of freedom with characters and names, and other packages take a advantage of that freedom too

For the record, I've never seen unicode used this extensively in any other Julia package's examples over the past ~two years of using Julia. I've also only seen a few examples of unicode being used in the API itself, and for those it was very minimal (e.g., a single operator like `∇`) and there were also ascii equivalents (noted earlier).

e.g., take a look at a random sampling of packages in, say, JuliaMath: [Interpolations.jl](https://github.com/JuliaMath/Interpolations.jl), [NFFT.jl](https://github.com/JuliaMath/NFFT.jl), [Combinatorics.jl](https://github.com/JuliaMath/Combinatorics.jl), [FFTW.jl](https://github.com/JuliaMath/FFTW.jl), [Polynomials.jl](https://github.com/JuliaMath/Polynomials.jl), [SpecialFunctions.jl](https://github.com/JuliaMath/SpecialFunctions.jl). Each of these packages _could_ choose to use unicode math extensively, but in actuality there is either no unicode, or extremely minimal use of unicode in both the examples and the API.

(Not that I think using other packages as examples is a good idea, though. There's always better ways to do things!)

---

At the end of the day though, it's up to you! Just wanted to give my honest feedback in the hopes of making this awesome package even better. Take from it as you will.

**Comment by tomchor** [comment]
I'm not sure how relevant this is gonna be, but I'm gonna leave this here in case it can help someone. In Linux you can easily(-ish) have your keyboard type unicode. There are a few ways listed [here](https://unix.stackexchange.com/a/487269). I do this (and set specific bindings for more common characters like ∂) and thus I rarely need to rely on any editor to type any of my code.

The easiest thing, however, is to follow what's described [here](https://superuser.com/a/28966) and add a greek layout to your keyboard, with a shortcut to change layouts. I do that too, and by pressing `Ctrl+Alt+k` I can swap back and forth between English and Greek, which allows me to type all the Greek letters effortlessly, not having to rely on any editor. (This obviously doesn't deal with all the other unicode characters, but it's a start!) I've only tried this on Manjaro Linux, but I'm pretty sure this method can be used on any OS.
